examples/text/plot_hashing_vs_dict_vectorizer.py

```diff
@@ -3,109 +3,343 @@
+# using the :func:`~sklearn.feature_extraction.DictVectorizer`. This is due to
+# hash collisions.
+#
+# The number of collisions can be reduced by increasing the feature space.
+# Notice that the speed of the vectorizer does not change significantly when
+# setting a large number of features, though it causes larger coefficient
+# dimensions and then requires more memory usage to store them, even if a
+# majority of them is inactive.
 
-print("FeatureHasher on frequency dicts")
 t0 = time()
-hasher = FeatureHasher(n_features=n_features)
+hasher = FeatureHasher(n_features=2**22)
 X = hasher.transform(token_freqs(d) for d in raw_data)
 duration = time() - t0
-print("done in %fs at %0.3fMB/s" % (duration, data_size_mb / duration))
-print("Found %d unique terms" % n_nonzero_columns(X))
-print()
 
-print("FeatureHasher on raw tokens")
+print(f"done in {duration:.3f}s at {data_size_mb / duration:.3f}MB/s")
+print(f"Found {n_nonzero_columns(X)} unique tokens")
+
+# %%
+# We confirm that the number of unique tokens gets closer to the number of
+# unique terms found by the :func:`~sklearn.feature_extraction.DictVectorizer`.
+#
+# **FeatureHasher on raw tokens**
+#
+# Alternatively, one can set `input_type="string"` in the
+# :func:`~sklearn.feature_extraction.FeatureHasher` to vectorize the strings
+# output directly from the customized `tokenize` function. This is equivalent to
+# passing a dictionary with an implied frequency of 1 for each feature name.
+
 t0 = time()
-hasher = FeatureHasher(n_features=n_features, input_type="string")
-X = hasher.transform(tokens(d) for d in raw_data)
+hasher = FeatureHasher(n_features=2**18, input_type="string")
+X = hasher.transform(tokenize(d) for d in raw_data)
 duration = time() - t0
-print("done in %fs at %0.3fMB/s" % (duration, data_size_mb / duration))
-print("Found %d unique terms" % n_nonzero_columns(X))
+dict_count_vectorizers["vectorizer"].append(
+    hasher.__class__.__name__ + "\non raw tokens"
+)
+dict_count_vectorizers["speed"].append(data_size_mb / duration)
+print(f"done in {duration:.3f}s at {data_size_mb / duration:.3f}MB/s")
+print(f"Found {n_nonzero_columns(X)} unique tokens")
+
+# %%
+# We now plot the speed of the above methods for vectorizing.
+
+import matplotlib.pyplot as plt
+
+fig, ax = plt.subplots(figsize=(12, 6))
+
+y_pos = np.arange(len(dict_count_vectorizers["vectorizer"]))
+ax.barh(y_pos, dict_count_vectorizers["speed"], align="center")
+ax.set_yticks(y_pos)
+ax.set_yticklabels(dict_count_vectorizers["vectorizer"])
+ax.invert_yaxis()
+_ = ax.set_xlabel("speed (MB/s)")
+
+# %%
+# In both cases :func:`~sklearn.feature_extraction.FeatureHasher` is
+# approximately twice as fast as
+# :func:`~sklearn.feature_extraction.DictVectorizer`. This is handy when dealing
+# with large amounts of data, with the downside of losing the invertibility of
+# the transformation, which in turn makes the interpretation of a model a more
+# complex task.
```

Comment:
```suggestion
# complex task.
#
# The `FeatureHeasher` with `input_type="string"` is slightly faster than
# the variant that works on frequency dict because it does not count repeated
# tokens: each token is implicitly counted once, even it was repeated. Depending
# on the downstream machine learning task, it can be a limitation or not.
```